x-pack/plugins/ecs_data_quality_dashboard/server/routes/results/get_results.ts

```diff
@@ -70,38 +64,33 @@ export const getResultsRoute = (
         }
 
         try {
-          // Confirm user has authorization for the requested patterns
-          const { patterns } = request.query;
-          const userEsClient = services.core.elasticsearch.client.asCurrentUser;
-          const privileges = await userEsClient.security.hasPrivileges({
-            index: [
-              { names: patterns.split(','), privileges: ['all', 'read', 'view_index_metadata'] },
-            ],
-          });
-          const authorizedPatterns = Object.keys(privileges.index).filter((pattern) =>
-            Object.values(privileges.index[pattern]).some((v) => v === true)
-          );
-          if (authorizedPatterns.length === 0) {
+          const { pattern } = request.query;
+
+          // Get authorized indices
+          const { client } = services.core.elasticsearch;
+          const indicesResponse = await client.asCurrentUser.indices.get({ index: pattern });
```

Comment:
### Summary

Consider the alternative implementation detailed below, because the following code:

```ts
        try {
          const { pattern } = request.query;

          // Get authorized indices
          const { client } = services.core.elasticsearch;
          const indicesResponse = await client.asCurrentUser.indices.get({ index: pattern });
```

1) Is not granular enough to ensure the user is authorized to access the specified indices, because it doesn't test the exact privileges

2) The `indicesResponse` is response is _massive_, because it includes the mappings of all indices matching the pattern. (These mappings, and most other entries in this response are not required.)

Thus, as a side effect of just visiting the Data Quality page, this very large response will be processed in the route handler, as demonstrated by adding the `console.log` below:

```ts
        try {
          const { pattern } = request.query;

          // Get authorized indices
          const { client } = services.core.elasticsearch;
          const indicesResponse = await client.asCurrentUser.indices.get({ index: pattern });

          console.log(
            `--> indicesResponse for pattern ${pattern}`,
            JSON.stringify(indicesResponse, null, 2)
          );

```

and then navigating to the Data Quality dashboard, as demonstrated in the following video:

<https://github.com/elastic/kibana/assets/4459398/74bc2758-27b1-49d0-97b8-5e6b8c47b4d9>

3) An `Error reading saved data quality check results` error toaster will be displayed for all users who don't have the privileges to check _every_ index in `indices`, as illustrated by the following screenshot:

![error-reading-saved-results](https://github.com/elastic/kibana/assets/4459398/ee99c038-a050-49dd-b6a0-0715f36abc8f)

because the following server-side error occurs, per the Kibana error log output below:

```
[2024-01-30T12:10:07.290-05:00][ERROR][plugins.ecsDataQualityDashboard] security_exception
        Root causes:
                security_exception: action [indices:admin/get] is unauthorized for user [read_but_not_view_index_metadata] with effective roles [read_but_not_view_index_metadata_role] on indices [auditbeat-*], this action is granted by the index privileges [view_index_metadata,manage,all]

```

For example, when a user who only has `monitor` and `read` (but not `view_index_metadata`) like the following role visits the data quality page:

![role](https://github.com/elastic/kibana/assets/4459398/5af1821e-a408-475f-905d-53ba10120d76)

That user (incorrectly) gets an error toaster, per the screenshot below:

![error_toasters](https://github.com/elastic/kibana/assets/4459398/055351c5-95ca-4daa-8d13-f9ab5ea9e7ea)

and errors like the following may appear in Kibana server log:

```
[2024-01-30T12:10:07.290-05:00][ERROR][plugins.ecsDataQualityDashboard] security_exception
        Root causes:
                security_exception: action [indices:admin/get] is unauthorized for user [read_but_not_view_index_metadata] with effective roles [read_but_not_view_index_metadata_role] on indices [auditbeat-*], this action is granted by the index privileges [view_index_metadata,manage,all]

```

### Proposed alternative implementation

The alternative implementation proposed below would

1) Update the `getResultsRoute` route (`/internal/ecs_data_quality_dashboard/results`) such that it received an array of `indexNames` in the `request body`, instead of a `pattern` like the current code today:

**Before:**

```ts
        try {
          const { pattern } = request.query;
```

**After:**

```ts
        try {
          const { indexNames } = request.query;
```

The contents of the `indexNames` will be provided by the caller of the API. The Data Quality dashboard should already have a valid collection of `indexNames` from the `stats` API.

Note that the user may not have all the privileges required to check all `indexNames` returned by `stats`. This is addressed by `2`, below.

2) We can't trust the input provided by the client, because any user invoking `/internal/ecs_data_quality_dashboard/results` could try guessing the name of an index, for example `users`, and then test for it's existance.

Consider removing the call to:

```ts
const indicesResponse = await client.asCurrentUser.indices.get({ index: pattern });
```

and replacing it with the following call to `userEsClient.security.hasPrivileges`:

```ts
          const privileges = await userEsClient.security.hasPrivileges({
            body: {
              index: [
                {
                  names: indexNames,
                  privileges: ['read', 'monitor', 'manage', 'view_index_metadata'],
                },
              ],
            },
          });
```

The `privileges` array above contains the full set of permissions required to check an index.

Next, with the following import:

```ts
import type { SecurityHasPrivilegesPrivileges } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
```

and the following helper functions:

```ts
export const hasRequiredPrivileges = ({
  manage = false,
  monitor = false,
  read = false,
  // eslint-disable-next-line @typescript-eslint/naming-convention
  view_index_metadata = false,
}: SecurityHasPrivilegesPrivileges): boolean => (monitor || manage) && read && view_index_metadata;

export type PrivilegesByIndex = Record<string, SecurityHasPrivilegesPrivileges>;

export const getAuthorizedIndices = (
  privilegesByIndex: PrivilegesByIndex | undefined
): string[] => {
  if (privilegesByIndex == null) {
    return [];
  }

  return Object.keys(privilegesByIndex).filter((index) =>
    hasRequiredPrivileges(privilegesByIndex[index])
  );
};
```

the `privileges` from:

```ts
          const privileges = await userEsClient.security.hasPrivileges({
            body: {
              index: [
                {
                  names: indexNames,
                  privileges: ['read', 'monitor', 'manage', 'view_index_metadata'],
                },
              ],
            },
          });
```

may be converted to `authorizedIndexNames` by calling the `getAuthorizedIndices` helper function:

```ts
const authorizedIndexNames = getAuthorizedIndices(privileges.index);
```

Putting all of the above together, it might look something like this:

```ts
export const hasRequiredPrivileges = ({
  manage = false,
  monitor = false,
  read = false,
  // eslint-disable-next-line @typescript-eslint/naming-convention
  view_index_metadata = false,
}: SecurityHasPrivilegesPrivileges): boolean => (monitor || manage) && read && view_index_metadata;

export type PrivilegesByIndex = Record<string, SecurityHasPrivilegesPrivileges>;

export const getAuthorizedIndices = (
  privilegesByIndex: PrivilegesByIndex | undefined
): string[] => {
  if (privilegesByIndex == null) {
    return [];
  }

  return Object.keys(privilegesByIndex).filter((index) =>
    hasRequiredPrivileges(privilegesByIndex[index])
  );
};

export const getResultsRoute = (
  router: IRouter<DataQualityDashboardRequestHandlerContext>,
  logger: Logger
) => {
  router.versioned
    .get({
      path: RESULTS_ROUTE_PATH,
      access: 'internal',
      options: { tags: ['access:securitySolution'] },
    })
    .addVersion(
      {
        version: INTERNAL_API_VERSION,
        validate: { request: { query: buildRouteValidation(GetResultQuery) } },
      },
      async (context, request, response) => {
        const services = await context.resolve(['core', 'dataQualityDashboard']);
        const resp = buildResponse(response);

        let index: string;
        try {
          index = await services.dataQualityDashboard.getResultsIndexName();
        } catch (err) {
          logger.error(`[GET results] Error retrieving results index name: ${err.message}`);
          return resp.error({
            body: `${API_RESULTS_INDEX_NOT_AVAILABLE}: ${err.message}`,
            statusCode: 503,
          });
        }

        try {
          const { indexNames } = request.query;

          // Get authorized indices
          const userEsClient = services.core.elasticsearch.client.asCurrentUser;
          const privileges = await userEsClient.security.hasPrivileges({
            body: {
              index: [
                {
                  names: indexNames,
                  privileges: ['read', 'monitor', 'manage', 'view_index_metadata'],
                },
              ],
            },
          });

          const authorizedIndexNames = getAuthorizedIndices(privileges.index);

          if (authorizedIndexNames.length === 0) {
            return response.ok({ body: [] });
          }

          // Get the latest result of each pattern
          const query = { index, ...getQuery(authorizedIndexNames) };

```

In conclusion, consider something like the alternative proposed above, because it:
- ensures the user has ALL the required privileges for each index
- does not trust input from the client
- resolves unexpected errors in the log / error toaster when the user does not have all required privileges
- is more performant, because the mappings for each index won't be returned to the route handler